### PR TITLE
Reduce build time by using no-restore

### PR DESCRIPTION
**Pull request by JunTaoLuo**
Addresses https://github.com/aspnet/AspNetCore/issues/11361

### CODE REVIEW

File: .azure/pipelines/ci.yml [hunk 1]
```diff
@@ -80,6 +80,7 @@ jobs:
     - script: ./build.cmd
             -ci
             -arch x86
+            -NoRestore
```
Comment (JunTaoLuo) [inline .azure/pipelines/ci.yml]: Can't do this one:
```
F:\workspace\_work\1\s\.dotnet\sdk\3.0.100-preview7-012821\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(234,5): error NETSDK1047: Assets file 'F:\workspace\_work\1\s\artifacts\obj\Microsoft.AspNetCore.App.Runtime\project.assets.json' doesn't have a target for '.NETCoreApp,Version=v3.0/win-x86'. Ensure that restore has run and that you have included 'netcoreapp3.0' in the TargetFrameworks for your project. You may also need to include 'win-x86' in your project's RuntimeIdentifiers. [F:\workspace\_work\1\s\src\Framework\src\Microsoft.AspNetCore.App.Runtime.csproj]
```

File: .azure/pipelines/ci.yml [hunk 2]
```diff
@@ -112,6 +112,7 @@ jobs:
     # Windows installers bundle both x86 and x64 assets
     - script: ./build.cmd
             -ci
+            -NoRestore
```
Comment (JunTaoLuo) [inline .azure/pipelines/ci.yml]: Can't do this either
```
F:\workspace\_work\1\s\src\Installers\Windows\AspNetCoreModule-Setup\ANCMIISExpressV2\AncmIISExpressV2.wixproj : error MSB4057: The target "Build" does not exist in the project.
```

**Comment by JunTaoLuo** [comment]
Though it's possible to remove the redundant restores, I'm seeing marginal benefit. For example, in this PR, I'm doing a restore before x64 build and skipping restore in BuildInstallers.
Comparing with this change: https://dev.azure.com/dnceng/public/_build/results?buildId=307019 and without https://dev.azure.com/dnceng/public/_build/results?buildId=307982.

It looks like the main difference is that we are splitting off the restore's time from x64 build into a separate step. I suspect this is because only the first time restore is doing a lot of work downloading artifacts and subsequent restores uses the cache and ends up being very fast.

My opinion is that this is a low value optimization. I'm okay with checking in this as is or closing. I don't want to continue the investigation on factoring out the restore in build site extensions for example.

Thoughts @dougbu @wtgodbe ?

**Comment by dougbu** [comment]
Is there any advantage to `--no-restore` in the Linux jobs that contain multiple steps e.g. 'Linux x64'?

I don't see the advantage of running `-restore -noBuild` then `-noRestore -build` in the Windows x64 case. I suspect there's a _disadvantage_ in depriving MSBuild of the information it gathered during the now-separate restore phase.

**Comment by JunTaoLuo** [comment]
I don't expect any gains in the linux build
1. I suspect the same will hold for linux, subsequent restores are fast
2. The build is already one of the shorter legs at ~23 mins so any optimization here won't be noticeable

**Comment by wtgodbe** [comment]
To make sure I'm understanding this correctly - the reason that we do x64 & x86 builds sequentially in the same job, is that the x86 build needs artifacts from the x64 build. The goal here was to allow all builds after the first one to skip restore, but we can't do that for x86 since x64 doesn't restore the right (x86) assets, and the gain is minimal for building the installers since that restore is fast anyways (maybe due to caching). If that's a more-or-less accurate summary, then I'm ok with abandoning this as the slight decrease in build time probably isn't worth the slight increase in complexity

**Comment by dougbu** [comment]
> I'm ok with abandoning this

Agreed. Let's close the issue and this PR.

**Pull request closed without merging**